origin/globus.go

```diff
@@ -0,0 +1,314 @@
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package origin
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/server_structs"
+	"github.com/pelicanplatform/pelican/server_utils"
+)
+
+type globusExportStatus string
+
+// For internal globusExports map
+type globusExport struct {
+	DisplayName      string             `json:"displayName"`
+	FederationPrefix string             `json:"federationPrefix"`
+	Status           globusExportStatus `json:"status"`
+	Description      string             `json:"description,omitempty"` // status description
+	HttpsServer      string             `json:"httpsServer"`           // server url to access files in the collection
+	Token            *oauth2.Token      `json:"-"`
+}
+
+// For UI
+type globusExportUI struct {
+	globusExport
+	UUID string `json:"uuid"`
+}
+
+const (
+	globusInactive  = "Inactive"
+	globusActivated = "Activated"
+)
+
+const globusTokenFileExt = ".tok" // File extension for caching Globus access token
+
+var (
+	globusExports      map[string]*globusExport // Key is the collection UUID
+	globusExportsMutex = sync.RWMutex{}
+)
+
+func InitGlobusBackend(exps []server_utils.OriginExport) error {
+	if server_utils.OriginStorageType(param.Origin_StorageType.GetString()) != server_utils.OriginStorageGlobus {
+		return fmt.Errorf("failed to initialize Globus backend: Origin.StorageType is not Globus: %s",
+			param.Origin_StorageType.GetString())
+	}
+	// Init map
+	globusExports = make(map[string]*globusExport)
+
+	// Check and setup token location
+	globusFdr := param.Origin_GlobusConfigLocation.GetString()
+	tokFdr := filepath.Join(globusFdr, "tokens")
+	if err := os.MkdirAll(tokFdr, 0755); err != nil {
+		return errors.Wrapf(err, "failed to create directory for Globus tokens: %s", tokFdr)
+	}
+
+	globusAuthCfg, err := GetGlobusOAuthCfg()
+	if err != nil {
+		return errors.Wrap(err, "failed to get Globus OAuth2 config")
+	}
+
+	// Set up Globus map
+	err = func() error {
```

Comment:
Reading through the code twice, I don't think I need to do this wrapper because the function body lasts until the end of the parent function. I'll remove the wrapper then.